### PR TITLE
feat(turborepo): web ui

### PR DESCRIPTION
**Pull request by NicholasLYang**
### Description

Implements a very very alpha version of a web UI. Uses a GraphQL query to get the current run and tasks.

### Testing Instructions

Would love some ideas here on tests.

To try out manually, go to #35066 in the other repo and run `turbo-studio`, then execute a turbo command with the `ui` flag set to `web`

### CODE REVIEW

File: crates/turbopack-cli/js/package.json [hunk 1]
```diff
@@ -8,7 +8,7 @@
     "check": "tsc --noEmit"
   },
   "dependencies": {
-    "@vercel/turbopack-ecmascript-runtime": "*"
+    "@vercel/turbopack-ecmascript-runtime": "workspace:*"
```
Comment (NicholasLYang) [inline crates/turbopack-cli/js/package.json:11]: For some reason this broke for me

File: crates/turborepo-ui/src/wui/mod.rs [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+                                    messages_to_send.push(Message::Text(message_payload));
+                                }
+
+                                for message in messages_to_send {
+                                    socket.send(message).await?;
+                                }
+                            }
+                        }
+                    } else {
+                        warn!("failed to deserialize message from client: {message_payload}");
```
Comment (chris-olszewski) [inline crates/turborepo-ui/src/wui/mod.rs:238]: I'm getting:
```
failed to deserialize message from client: {"type":"Ack","payload":{}}
```

File: crates/turborepo-ui/src/wui/mod.rs [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+}
+
+async fn handle_socket_inner(mut socket: WebSocket, state: AppState) -> Result<(), Error> {
+    let mut state = state.clone();
+    let mut acks = HashSet::new();
+    let mut interval = tokio::time::interval(std::time::Duration::from_millis(100));
+    'socket_loop: loop {
+        select! {
+            biased;
+            Ok(event) = state.rx.recv() => {
```
Comment (chris-olszewski) [inline crates/turborepo-ui/src/wui/mod.rs:185]: I think we need to handle the case where the Web UI ends up lagging.

File: crates/turborepo-ui/src/wui/mod.rs [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+                for (event, id) in messages.borrow().iter() {
+                    if !acks.contains(id) {
+                        let message_payload = serde_json::to_string(event).unwrap();
+                        messages_to_send.push(Message::Text(message_payload));
+                    }
+                };
+
+                for message in messages_to_send {
+                    socket.send(message).await?;
+                }
```
Comment (chris-olszewski) [inline crates/turborepo-ui/src/wui/mod.rs:209]: We should drop the lock before we do a serial send.

File: crates/turborepo-ui/src/wui/mod.rs [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+    Ack { id: u32 },
+    /// Asks for all messages from the given id onwards
+    CatchUp { start_id: u32 },
+}
+
+struct AppState {
+    rx: tokio::sync::broadcast::Receiver<WebUIEvent>,
+    // We use a tokio::sync::Mutex here because we want this future to be Send.
+    #[allow(clippy::type_complexity)]
+    messages: Arc<Mutex<RefCell<Vec<(WebUIEvent, u32)>>>>,
```
Comment (chris-olszewski) [inline crates/turborepo-ui/src/wui/mod.rs:154]: I think we really want to use a new tokio task to manage `messages` state. If the socket is slow on receiving these, we can end up dropping a lot of messages due to how `tokio::broadcast` works.

We're doing a lot of sends on the socket handler task and if one of them ends up being slow we stop moving forward.

**Review by chris-olszewski — CHANGES_REQUESTED** [review]
This is awesome. I think we just need to make the refactor the socket handler a little bit so it can continue processing events if a socket ends being slow.

File: crates/turborepo-lib/src/commands/run.rs [hunk 6]
```diff
@@ -44,15 +44,15 @@ pub async fn run(base: CommandBase, telemetry: CommandEventBuilder) -> Result<i3
...
+        let (sender, handle) = run.start_ui()?.unzip();
+
         let result = run.run(sender.clone(), false).await;
 
         if let Some(analytics_handle) = analytics_handle {
             analytics_handle.close_with_timeout().await;
         }
 
-        if let (Some(handle), Some(sender)) = (handle, sender) {
-            sender.stop();
```
[the remark this answers is not in the document]
Comment (NicholasLYang) [reply crates/turborepo-lib/src/commands/run.rs:55]: Kinda. I wanna keep this open even after a run for the web ui. I'll make it only stop the TUI

**Comment by NicholasLYang** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/vercel/turborepo/8895?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#8895** <a href="https://app.graphite.dev/github/pr/vercel/turborepo/8895?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#8977** <a href="https://app.graphite.dev/github/pr/vercel/turborepo/8977?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @NicholasLYang and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: crates/turborepo-ui/src/wui/server.rs [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+            .endpoint("/")
+            .subscription_endpoint("/subscriptions")
+            .finish(),
+    )
+}
+
+pub async fn start_server(
+    rx: tokio::sync::mpsc::UnboundedReceiver<WebUIEvent>,
+) -> Result<(), crate::Error> {
+    let state = Arc::new(Mutex::new(RefCell::new(WebUIState::default())));
```
Comment (chris-olszewski) [inline crates/turborepo-ui/src/wui/server.rs:74]: I'm not sure if I understand the use of `RefCell` inside of the `Mutex`. Doesn't the `Mutex` already ensure that there's only ever one access to the underlying data at a time? Why are dynamic borrows necessary inside of the restriction?

If this is necessary a quick comment explaining why would be very appreciated.

**Review by chris-olszewski — CHANGES_REQUESTED** [review]
Looks like tests are failing.

Aside from that just a minor comment on use of `RefCell` inside of `Mutex`.

**Review by chris-olszewski — APPROVED** [review]

**Pull request merged**